# Load libraries and user functions ----------------------------------------------------------

ipak <- function(pkg){
  new.pkg <- pkg[!(pkg %in% installed.packages()[, "Package"])]
  # if (length(new.pkg))
  #   install.packages(new.pkg, dependencies = TRUE)
  sapply(pkg, library, character.only = TRUE)
}

packages<-c("Seurat", "dplyr", "colorRamps")

# Load libraries
ipak(packages)

# Define variables --------------------------------------------------------

all_data.files <- c("WTLSK" = "WT_LSK/outs/filtered_gene_bc_matrices/mm10/",
                    "HETLSK" = "HET_LSK/outs/filtered_gene_bc_matrices/mm10/",
                    "WTB220" = "WT_B220/outs/filtered_gene_bc_matrices/mm10/",
                    "HETB220" = "HET_B220/outs/filtered_gene_bc_matrices/mm10/")
ProjectName<-"Stjude_cca"
genome<- "mm10"
Run_mito_filter = FALSE
output_prefix <-"20180601_Stjude_cca"
max_pcs <-21
resolution_list <-c(0.6, 0.8, 1.0, 1.2, 1.5, 2.0, 2.5)
#resolution_list<-0.6
# Define basic color palette ----------------------------------------------

my_palette<-c("#cb4bbe",
                 "lightskyblue",
                 "grey37",
                 "#53ba48",
                 "moccasin",
                 "#dcca44",
                 "#502d70",
                 "#afd671",
                 "#cb4471",
                 "#69da99",
                 "#d44732",
                 "#6fd3cf",
                 "#5d2539",
                 "#cfd0a0",
                 "blue",
                 "#d2883b",
                 "#6793c0",
                 "#898938",
                 "#c98ac2",
                 "yellow",
                 "#c4c0cc",
                 "#7d3d23",
                 "#00a5ff",
                 "#d68d7a",
                 "#a2c1a3")

# Load data ---------------------------------------------------------------

# Alter output name based on filter parameters

if(Run_mito_filter == TRUE){
  output_prefix<-paste(output_prefix, "_mt", sep = "")
}

multi_object_list<-list()
list_of_names<-names(all_data.files)
list_of_names


# Create multi_object_list
create_multi_object_list<-function(x){
  my_object.data<-Read10X(x)
  my_object<- CreateSeuratObject(raw.data = my_object.data, min.cells = 3, min.genes = 200, project = ProjectName)
  
  # Apply mito.filter if applicable
  if(Run_mito_filter == TRUE){
    mito.genes<-grep(pattern = "^MT-", x = rownames(x=my_object@data), value = TRUE, ignore.case = TRUE)
    percent.mito <-Matrix::colSums(my_object@raw.data[mito.genes,])/Matrix::colSums(my_object@raw.data)
    my_object<-AddMetaData(my_object, metadata = percent.mito, col.name = "percent.mito")
  } else {
    print("Not running mito filter")
  }
  my_object<-NormalizeData(my_object, normalization.method = "LogNormalize", scale.factor = 10000)
  my_object<-FindVariableGenes(my_object,mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)

  # Scale on percent.mito if applicable
  if(Run_mito_filter == TRUE){
    my_object<-ScaleData(my_object, vars.to.regress = c("nUMI", "percent.mito"))
  } else {
    my_object<-ScaleData(my_object, vars.to.regress = "nUMI")
  }
  my_object@meta.data$sample<-names(x)
  multi_object_list[[names(x)]]<-my_object
}

multi_object_list<-lapply(1:length(all_data.files), function(x) create_multi_object_list(all_data.files[x]))
names(multi_object_list)
names(multi_object_list)<-list_of_names
multi_object_list

genes.use<-c()
for (i in 1:length(multi_object_list)){
  print(i)
  if(length(multi_object_list[[i]]@var.genes)>=1000){
  genes.use<-c(genes.use, head(rownames(multi_object_list[[i]]@hvg.info), 1000))
  } else {
    genes.use<-c(genes.use, head(rownames(multi_object_list[[i]]@hvg.info)))
  }
}

genes.use<-names(which(table(genes.use)>1))
for (i in 1:length(multi_object_list)){
  genes.use<-genes.use[genes.use %in% rownames(multi_object_list[[i]]@scale.data)]
}
head(genes.use)
# Run multi-set CCA


# Iterate across multiple resolutions -------------------------------------


integrated_object<-RunMultiCCA(multi_object_list, genes.use = genes.use, num.ccs = max_pcs)

png(filename = paste(output_prefix,"_dim",max_pcs, "_BicorPlot.png", sep = ""), height = 800, width = 800)
MetageneBicorPlot(integrated_object, grouping.var = "sample", dims.eval = 1:max_pcs)
dev.off()

integrated_object<-CalcVarExpRatio(integrated_object, reduction.type = "pca", grouping.var = "sample", dims.use = 1:max_pcs)
integrated_object<-SubsetData(integrated_object, subset.name = "var.ratio.pca", accept.low = 0.5)
integrated_object<-AlignSubspace(integrated_object, reduction.type = "cca", grouping.var = "sample", dims.align = 1:max_pcs)

  
for(my_resolution in resolution_list){
  integrated_object<-FindClusters(integrated_object, reduction.type = "cca.aligned", dims.use = 1:max_pcs, save.SNN = T, resolution = my_resolution)
  integrated_dba<-RunTSNE(integrated_object, reduction.use = "cca.aligned", dims.use = 1:max_pcs)
  
  # Create color pallete
  new_length <-0
  if(length(unique(integrated_dba@ident)) > length(my_palette)){
    new_length<-length(unique(integrated_dba@ident)) - length(my_palette)
  }
  new_length
  my_palette<-c(my_palette, primary.colors(new_length))
  

  png(filename = paste(output_prefix, "dim",max_pcs, "res",resolution,"_tsne.png", sep = ""), width = 800, height = 800)
  TSNEPlot(integrated_dba, do.label = TRUE, colors.use = my_palette, label.size = 10)
  dev.off()

  png(filename = paste(output_prefix, "dim",max_pcs, "res",resolution,"_tsne-Labeled.png", sep = ""), width = 800, height = 800)
  TSNEPlot(integrated_dba, do.label = TRUE, group.by = "sample", colors.use = my_palette, label.size = 10)
  dev.off()

  saveRDS(integrated_dba, file = paste(output_prefix, "_dim", max_pcs, "res", my_resolution, "_tsne.rds", sep = ""))
  
  # Tabulate data -----------------------------------------------------------
  
  #Number of cells in each cluster
  table(integrated_dba@ident)
  write.table(table(integrated_dba@ident), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_popCounts.txt", sep = ""), row.names = FALSE, quote = FALSE, sep = "\t")
  
  
  #Number of cells in each ID
  table(integrated_dba@meta.data$orig.ident)
  write.table(table(integrated_dba@meta.data$orig.ident), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_popCounts.txt", sep = ""), row.names = FALSE, quote = FALSE, sep = "\t", append = TRUE)
  
  #Proportion of cells in each cluster
  prop.table(x = table(integrated_dba@ident))
  write.table(prop.table(x = table(integrated_dba@ident)), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_popCounts.txt", sep = ""), row.names = FALSE, quote = FALSE, sep = "\t", append = TRUE)
  
  #Number of cells from each ID in each cluster
  table(integrated_dba@ident, integrated_dba@meta.data$orig.ident)
  write.table(table(integrated_dba@ident, integrated_dba@meta.data$orig.ident), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_popCounts.txt", sep = ""), row.names = FALSE, quote = FALSE, sep = "\t", append = TRUE)
  
  #Proportion of cells of each ID in each cluster
  prop.table(x = table(integrated_dba@ident, integrated_dba@meta.data$orig.ident))
  write.table(prop.table(x = table(integrated_dba@ident, integrated_dba@meta.data$orig.ident)), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_popCounts.txt", sep = ""), row.names = FALSE, quote = FALSE, sep = "\t", append = TRUE)
  
  #Average expression of each gene in each cluster
  avgexp<-AverageExpression(integrated_dba)
  write.table(table(integrated_dba@ident), file = paste(output_prefix, "dim",max_pcs, "res",resolution,"_AvgXprsn.txt", sep = ""), row.names = TRUE, quote = FALSE, sep = "\t")


}


q()
































